jobsdb/jobsdb.go

```diff
@@ -1133,27 +1132,43 @@ func (jd *HandleT) migrateJobs(srcDS dataSetT, destDS dataSetT) (noJobsMigrated
 	jd.assertError(err)
 	jobsToMigrate := append(unprocessedList, retryList...)
 	noJobsMigrated = len(jobsToMigrate)
-	//Copy the jobs over. Second parameter (true) makes sure job_id is copied over
-	//instead of getting auto-assigned
-	err = jd.storeJobsDS(destDS, true, jobsToMigrate) //TODO: switch to transaction
-	jd.assertError(err)
 
-	//Now copy over the latest status of the unfinished jobs
 	var statusList []*JobStatusT
 	for _, job := range retryList {
-		newStatus := JobStatusT{
-			JobID:         job.JobID,
-			JobState:      job.LastJobStatus.JobState,
-			AttemptNum:    job.LastJobStatus.AttemptNum,
-			ExecTime:      job.LastJobStatus.ExecTime,
-			RetryTime:     job.LastJobStatus.RetryTime,
-			ErrorCode:     job.LastJobStatus.ErrorCode,
-			ErrorResponse: job.LastJobStatus.ErrorResponse,
-		}
-		statusList = append(statusList, &newStatus)
+		job.LastJobStatus.JobID = job.JobID
+		statusList = append(statusList, &job.LastJobStatus)
 	}
-	err = jd.updateJobStatusDS(destDS, statusList, []string{}, nil) //TODO: switch to transaction
+
+	err = jd.writeJobsAndStatusesToDS(destDS, jobsToMigrate, statusList)
 	jd.assertError(err)
+	return
+}
+
+func (jd *HandleT) writeJobsAndStatusesToDS(destDS dataSetT, jobsToMigrate []*JobT, statusList []*JobStatusT) (err error) {
+	txn, err := jd.dbHandle.Begin()
```

Comment:
While we are at it, we should consider moving rename/drop that we do in `postMigrateHandleDS` within this transaction. That would avoid the scenario where there is a crash right before `jd.JournalMarkStart(postMigrateDSOperation...`